### ISSUE TITLE
Bug when DesktopInstanceField.HasSimpleValue is called on a Int32[] field

### ISSUE THREAD

**Issue by chrisnas** [microsoft/clrmd#108]
While writing the ClrMD-based code to list the elements of a **ConcurrentQueue<int>**, it seems that **HasSimpleValue** is triggering a buggy state.
`    var t = heap.GetObjectType(0x00000145071476f0);  // --> System.Collections.Concurrent.ConcurrentQueue+Segment<System.Int32>
`
this segment has a int[] field called **m_array** --> we need to figure out if it is an array of simple values:
```
   var f = t.GetFieldByName("m_array");
   var r = f.HasSimpleValue;
```
When **f.HasSimpleValue** is called, the underlying type gets created and **GetTypeByMethodTable** is called with 0/null as its **obj** parameter: a new **DesktopHeapType** instance is added into **_types** (+ its **mt** into **_indices**) with a **ComponentType** set to **null**.
 	

> Microsoft.Diagnostics.Runtime.dll!Microsoft.Diagnostics.Runtime.Desktop.V46GCHeap.GetTypeByMethodTable(ulong mt, ulong _, ulong obj) Line 2294	C# 	Microsoft.Diagnostics.Runtime.dll!Microsoft.Diagnostics.Runtime.Desktop.DesktopModule.EnumerateTypes() Line 167	C# 	Microsoft.Diagnostics.Runtime.dll!Microsoft.Diagnostics.Runtime.Desktop.DesktopGCHeap.InitBasicTypes() Line 1056	C# 	Microsoft.Diagnostics.Runtime.dll!Microsoft.Diagnostics.Runtime.Desktop.DesktopGCHeap.GetBasicType(Microsoft.Diagnostics.Runtime.ClrElementType elType) Line 1010	C# 	Microsoft.Diagnostics.Runtime.dll!Microsoft.Diagnostics.Runtime.Desktop.DesktopInstanceField.GetType(Microsoft.Diagnostics.Runtime.Desktop.DesktopGCHeap heap, Microsoft.Diagnostics.Runtime.Desktop.IFieldData data, System.IntPtr sig, int sigLen, Microsoft.Diagnostics.Runtime.ClrElementType elementType) Line 617	C# 	Microsoft.Diagnostics.Runtime.dll!Microsoft.Diagnostics.Runtime.Desktop.DesktopInstanceField..ctor.AnonymousMethod__0() Line 487	C# 	mscorlib.dll!System.Lazy<Microsoft.Diagnostics.Runtime.Desktop.BaseDesktopHeapType>.CreateValue() Line 437	C# 	mscorlib.dll!System.Lazy<Microsoft.Diagnostics.Runtime.Desktop.BaseDesktopHeapType>.LazyInitValue() Line 388	C# 	Microsoft.Diagnostics.Runtime.dll!Microsoft.Diagnostics.Runtime.Desktop.DesktopInstanceField.ElementType.get() Line 637	C# 	Microsoft.Diagnostics.Runtime.dll!Microsoft.Diagnostics.Runtime.Desktop.DesktopInstanceField.HasSimpleValue.get() Line 667	C#

Next, **HasSimpleValue** tries to get the **ElementType**:
```
public override ClrElementType ElementType
{
    get
    {
        if (_elementType != ClrElementType.Unknown)
            return _elementType;

        ClrType type = _type.Value;
```
and the **_type.Value** triggers the lazy initialization of the corresponding type that ends up to the **GetType** method:

> Microsoft.Diagnostics.Runtime.dll!Microsoft.Diagnostics.Runtime.Desktop.DesktopInstanceField.GetType(Microsoft.Diagnostics.Runtime.Desktop.DesktopGCHeap heap, Microsoft.Diagnostics.Runtime.Desktop.IFieldData data, System.IntPtr sig, int sigLen, Microsoft.Diagnostics.Runtime.ClrElementType elementType) Line 619	C# 	Microsoft.Diagnostics.Runtime.dll!Microsoft.Diagnostics.Runtime.Desktop.DesktopInstanceField..ctor.AnonymousMethod__0() Line 487	C# 	mscorlib.dll!System.Lazy<Microsoft.Diagnostics.Runtime.Desktop.BaseDesktopHeapType>.CreateValue() Line 437	C# 	mscorlib.dll!System.Lazy<Microsoft.Diagnostics.Runtime.Desktop.BaseDesktopHeapType>.LazyInitValue() Line 388	C# 	Microsoft.Diagnostics.Runtime.dll!Microsoft.Diagnostics.Runtime.Desktop.DesktopInstanceField.ElementType.get() Line 637	C# 	Microsoft.Diagnostics.Runtime.dll!Microsoft.Diagnostics.Runtime.Desktop.DesktopInstanceField.HasSimpleValue.get() Line 667	C#

where **ComponentType** for Int32[] type is null and **GetBasicType()** returns **null**: the **null** **ComponentType** becomes **heap.ErrorType**
```
else if (result.ComponentType == null)
    if ((result.ComponentType = heap.GetBasicType((ClrElementType)etype)) == null)
        result.ComponentType = heap.ErrorType;
```

From that point, it seems that result (i.e. type for Int32[]) is now in the ClrMD cache with an invalid **ComponentType** field.
When we try to get the reference to the **m_array** field and get its **ClrType**, we get an error:
```
   var v = f.GetValue(0x00000145071476f0);
   var tt = heap.GetObjectType((ulong) v);
   Console.WriteLine(tt.ComponentType.ToString());  // --> ERROR
```

One possible fix could be to reset the **ComponentType** to **null** in **DesktopInstanceField.GetType()** if **heap.GetBasicType()** returns null instead of erro (it worked for us) or don't add the type in the ClrMD cache in **V46GCHeap.GetTypeByMethodTable()** if it is an array + obj == 0 + ComponentType is null

REPRO: write a console application that instanciate a ConcurrentQueue<int> and dump it with procdump -ma. Get the address of the instance in WinDBG with 

> 0:000>!dumpheap -stat
> 00007ffbcb045740        2           96 System.Collections.Concurrent.ConcurrentQueue`1[[System.Int32, mscorlib]]
> 0:000> !DumpHeap /d -mt 00007ffbcb045740
>          Address               MT     Size
> 0000014507147548 00007ffbcb045740       48     
> **00000145071476c0** 00007ffbcb045740       48     
> 

get the address of a segment:

> 0:000> !DumpObj /d **00000145071476c0**
> Name:        System.Collections.Concurrent.ConcurrentQueue`1[[System.Int32, mscorlib]]
> MethodTable: 00007ffbcb045740
> EEClass:     00007ffbcb0407d8
> Size:        48(0x30) bytes
> File:        C:\WINDOWS\Microsoft.Net\assembly\GAC_64\mscorlib\v4.0_4.0.0.0__b77a5c561934e089\mscorlib.dll
> Fields:
>               MT    Field   Offset                 Type VT     Attr            Value Name
> 00007ffbcb046a20  4001849        8 ...Int32, mscorlib]]  0 instance **00000145071476f0** m_head

and run the code shown earlier in this description:
```
  var t = heap.GetObjectType(0x00000145071476f0);
  var f = t.GetFieldByName("m_array");
  var r = f.HasSimpleValue;
  var v = f.GetValue(0x00000145071476f0);
  var tt = heap.GetObjectType((ulong) v);
  Console.WriteLine(tt.ComponentType.ToString());  // --> ERROR
```

**Comment by leculver** [comment]
Thank you for the detailed bug report!

I have some vague plans to revisit a lot of the internals of the ClrMD type system in ClrMD 1.1 or 1.2 (which is a while out).  One of the goals of that planned work is to simplify and rework a lot of the internal parts of ClrMD, especially around some of the really bad heuristics when we have to start working with ClrElementType.

I'm going to delay looking into/fixing this bug until I either do that work, or decide that I won't go ahead with a rewrite.  (In the case where I don't rewrite some of the internals I will still tackle this bug as a one-off fix.)

**Comment by leculver** [comment]
The updates to this codepath in ClrMD 2.0 fixes this problem, I've added tests in PR #363 which tests this scenario directly to ensure it doesn't regress.  Unfortunately I don't plan to fix this in ClrMD 1.1 because the "fix" is just me rewriting most of the entire system...  If you need this fixed in 1.1 I'd accept a pull request if you made one with your proposed fix.

For the upcoming .Net Core 5, I hope to add some additional dac private interface functions to avoid this kind of problem entirely by reporting the actual component TypeHandle for an array type...  This should avoid needing to have an object to get the component.  If that change goes into .Net Core 5 I'll be updating 2.0 and 1.1 to support it...though it would only help when debugging a .Net Core 5 and beyond runtime.
